README.md
[![Coverage Status](https://coveralls.io/repos/github/Peterfurax/flatplan_api/badge.svg?branch=master)](https://coveralls.io/github/Peterfurax/flatplan_api?branch=master) [![travis](https://travis-ci.org/Peterfurax/flatplan_api.svg?branch=master)](https://travis-ci.org/Peterfurax/flatplan_api) [![Known Vulnerabilities](https://snyk.io/test/github/peterfurax/flatplan_api/badge.svg)](https://snyk.io/test/github/peterfurax/flatplan_api) [![Dependency Status](https://david-dm.org/peterfurax/flatPlan_api.svg)](https://david-dm.org/peterfurax/flatPlan_api) [![macOS Build Status](https://circleci.com/gh/Peterfurax/flatplan_api.svg?style=shield)](https://circleci.com/gh/Peterfurax/flatPlan_api) [![Windows Build status](https://ci.appveyor.com/api/projects/status/github/Peterfurax/flatPlan_api?svg=true&passingText=build%20-%20OK)](https://ci.appveyor.com/project/jshint/jshint/branch/master) [![bitHound Code](https://www.bithound.io/github/Peterfurax/flatplan_api/badges/code.svg)](https://www.bithound.io/github/Peterfurax/flatplan_api)[![bitHound Dev Dependencies](https://www.bithound.io/github/Peterfurax/flatplan_api/badges/devDependencies.svg)](https://www.bithound.io/github/Peterfurax/flatplan_api/master/dependencies/npm)

# flatplan_api

1.0.1


CHANGELOG.md
[![Coverage Status](https://coveralls.io/repos/github/Peterfurax/flatplan_api/badge.svg?branch=master)](https://coveralls.io/github/Peterfurax/flatplan_api?branch=master) [![travis](https://travis-ci.org/Peterfurax/flatplan_api.svg?branch=master)](https://travis-ci.org/Peterfurax/flatplan_api) [![Known Vulnerabilities](https://snyk.io/test/github/peterfurax/flatplan_api/badge.svg)](https://snyk.io/test/github/peterfurax/flatplan_api) [![Dependency Status](https://david-dm.org/peterfurax/flatPlan_api.svg)](https://david-dm.org/peterfurax/flatPlan_api) [![macOS Build Status](https://circleci.com/gh/Peterfurax/flatplan_api.svg?style=shield)](https://circleci.com/gh/Peterfurax/flatPlan_api) [![Windows Build status](https://ci.appveyor.com/api/projects/status/github/Peterfurax/flatPlan_api?svg=true&passingText=build%20-%20OK)](https://ci.appveyor.com/project/jshint/jshint/branch/master)

# 1.0.3 put option


test/test.js
"use strict";
/*jslint node: true */
/*jshint esversion: 6 */
/*jshint expr: true */
/*jshint loopfunc: true */
// Require the dev-dependencies
const chai = require('chai');
const chaiHttp = require('chai-http');
const server = require('../dist/server.js');
const getTest = require('./provider/getTest');
const getTestErr = require('./provider/getTestErr');
const should = chai.should();
let logData = [];
let counter = 0;
const testGetArr = getTest;
const testGetArrErr = getTestErr;

function logger(mess) {
  counter += 1;
  logData.push({
    counter: mess
  });
  console.log(counter);
  // console.log(logData);
}

function statusHttp(uri, httpReponseCode) {
  return new Promise((resolve, reject) => {
    if (typeof httpReponseCode !== 'number') {
      reject('http code must be a number');
    }
    it('Http status res ' + httpReponseCode + '', (done) => {
      chai.request(server)
        .get(uri)
        .end((err, res) => {
          resolve(res.should.have.status(httpReponseCode));
          done();
        });
    });
  });
}

function putT(uri) {
  return new Promise((resolve, reject) => {
    it('should be json', (done) => {
      chai.request(server)
        .put(uri)
        .end((err, res) => {
          if (err) {
            reject(err);
          }
          resolve(res.should.be.json);
          done();
        });
    });
  });
}

function beJson(uri) {
  return new Promise((resolve, reject) => {
    it('should be json', (done) => {
      chai.request(server)
        .get(uri)
        .end((err, res) => {
          if (err) {
            reject(err);
          }
          resolve(res.should.be.json);
          done();
        });
    });
  });
}

function beObject(uri) {
  return new Promise((resolve, reject) => {
    it('should be objet', (done) => {
      chai.request(server)
        .get(uri)
        .end((err, res) => {
          if (err) {
            reject(err);
          }
          resolve(res.body.should.be.a('object'));
          done();
        });
    });
  });
}

function keyIsString(uri) {
  return new Promise((resolve, reject) => {
    it('should body.message is string', (done) => {
      chai.request(server)
        .get(uri)
        .end((err, res) => {
          if (err) {
            reject(err);
          }
          resolve(res.body.message.should.be.a('string'));
          done();
        });
    });
  });
}

function bodyHaveProperty(uri, waitProperty) {
  return new Promise((resolve, reject) => {
    it('should have property ' + waitProperty + '', (done) => {
      chai.request(server)
        .get(uri)
        .end((err, res) => {
          if (err) {
            reject(err);
          }
          resolve(res.body.should.have.property(waitProperty));
          done();
        });
    });
  });
}

function stringIsEqual(uri, waitingMessage) {
  return new Promise((resolve, reject) => {
    it('should message equal = ' + waitingMessage + '', (done) => {
      chai.request(server)
        .get(uri)
        .end((err, res) => {
          if (err) {
            reject(err);
          }
          resolve(res.body.message.should.equal(waitingMessage));
          done();
        });
    });
  });
}
const httpStatus = 200;
const httpStatusErr = 400;
const httpStatusDrop = 404;
const welcomeMessage = 'Welcome on flatplan_api !';
const testPutArr = [
'/api/produit/sli',
'/api/produit/sli/parution/20160101',
'/api/produit/sli/parution/20160101/folio/01',
'/api/produit/sli/parution/20160101/folio/01/status/newStatus'
];

function putTest() {
  for (var i = 0; i < testPutArr.length; i++) {
    const uri = testPutArr[i];
    describe(uri, () => {
      Promise.all([
        putT(uri)
      ])
        .then(result => logger(result), err => logger(err))
        .catch(err => logger(err));
    });
  }
}

function getTestArr() {
  for (var i = 0; i < testGetArr.length; i++) {
    const uri = testGetArr[i];
    describe(uri, () => {
      Promise.all([
        statusHttp(uri, httpStatus),
        beJson(uri)
      ])
        .then(result => logger(result), err => logger(err))
        .catch(err => logger(err));
    });
  }
}

function getTestErrArr() {
  for (var i = 0; i < testGetArrErr.length; i++) {
    const uri = testGetArrErr[i];
    describe(uri, () => {
      Promise.all([
        statusHttp(uri, httpStatusErr),
        beJson(uri)
      ])
        .then(result => logger(result), err => logger(err))
        .catch(err => logger(err));
    });
  }
}
// During the test the env variable is set to test
process.env.NODE_ENV = 'test';
// Chai use
chai.use(chaiHttp);
describe('Api response ', () => {
  describe('#PATH', () => {
    describe('#/', () => {
      statusHttp('/', httpStatusDrop)
        .then(result => logger(result), err => logger(err))
        .catch(err => logger(err));
    });
    describe('#/api', () => {
      const uri = '/api';
      const waitRep = 'message';
      Promise.all([
        statusHttp(uri, httpStatus),
        beJson(uri),
        beObject(uri),
        keyIsString(uri),
        bodyHaveProperty(uri, waitRep),
        stringIsEqual(uri, welcomeMessage)
      ])
        .then(result => logger(result), err => logger(err))
        .catch(err => logger(err));
    });
    Promise.all([
      getTestArr(),
      getTestErrArr(),
      putTest()
    ])
      .then(result => logger(result), err => logger(err))
      .catch(err => logger(err));
  });
});


test/provider/getTest.js
// INIT FILE SETUP
// =============================================================================
// - use strict option
"use strict";
let getTest = [];
// - Define product array you want to test ["/sli", "/ewe",...]
const productArr = ["/sli", "/ewe"];
// - Define parution array you want to test ["/20160101", "/20160102",...]
const parutionArr = ["/20160101", "/20160102"];
// - Define folio array you want to test ["/01", "/02",...]
const folioArr = ["/01"];
// - Origin path
const oriPath = "/api";
// - Product path
const produitPath = "/produit";
// - Parution path
const parutionPath = "/parution";
// - Folio path
const folioPath = "/folio";
// - Status path
const statusPath = "/status";
// - Alpha path
const alphaPath = oriPath + produitPath;
// FUNCTION
// =============================================================================
/**
 * [return a Listing test path]
 * @method makeListing
 * @return {Object}    ['stringPath',...]
 */
function makeListing() {
  // API WELCOME
  // -----
  // '/api/'
  getTest.push(alphaPath);
  for (let a = 0; a < productArr.length; a++) {
    // PRODUIT
    // -----
    // '/api/produit/sli'
    getTest.push(alphaPath + productArr[a]);
    for (let b = 0; b < parutionArr.length; b++) {
      // PARUTION
      // -----
      // '/api/produit/sli/parution'
      getTest.push(alphaPath + productArr[a] + parutionPath);
      getTest.push(alphaPath + productArr[a] + parutionPath + parutionArr[b]);
      for (let c = 0; c < folioArr.length; c++) {
        // FOLIOS
        // -----
        // '/api/produit/sli/parution/ERR/folio'
        getTest.push(alphaPath + productArr[a] + parutionPath + parutionArr[b] + folioPath);
        getTest.push(alphaPath + productArr[a] + parutionPath + parutionArr[b] + folioPath + folioArr[c]);
        getTest.push(alphaPath + productArr[a] + parutionPath + parutionArr[b] + folioPath + folioArr[c] + statusPath);
      }
    }
  }
}
makeListing();
// - console.log(getTest);
// -  [ '/api/produit',
//   '/api/produit/sli',
//   '/api/produit/sli/parution',
//   '/api/produit/sli/parution/20160101',
//   '/api/produit/sli/parution/20160101/folio/01',
//   '/api/produit/sli/parution/20160101/folio/01/status',
//   '/api/produit/sli/parution/20160101/folio/02',
//   '/api/produit/sli/parution/20160101/folio/02/status',
//   '/api/produit/sli/parution',
//   '/api/produit/sli/parution/20160102',
//   '/api/produit/sli/parution/20160102/folio/01',
//   '/api/produit/sli/parution/20160102/folio/01/status',
//   '/api/produit/sli/parution/20160102/folio/02',
//   '/api/produit/sli/parution/20160102/folio/02/status',
//   '/api/produit/ewe',
//   '/api/produit/ewe/parution',
//   '/api/produit/ewe/parution/20160101',
//   '/api/produit/ewe/parution/20160101/folio/01',
//   '/api/produit/ewe/parution/20160101/folio/01/status',
//   '/api/produit/ewe/parution/20160101/folio/02',
//   '/api/produit/ewe/parution/20160101/folio/02/status',
//   '/api/produit/ewe/parution',
//   '/api/produit/ewe/parution/20160102',
//   '/api/produit/ewe/parution/20160102/folio/01',
//   '/api/produit/ewe/parution/20160102/folio/01/status',
//   '/api/produit/ewe/parution/20160102/folio/02',
//   '/api/produit/ewe/parution/20160102/folio/02/status' ]
module.exports = getTest;


test/provider/apiListErr.js
"use strict";
let apiList = [];
const errTest = ["/err"];
const productArr = ["/sli", "/ewe"];
const parutionArr = ["/20160101", "/20160102"];
const folioArr = ["/01", "/02"];
const oriPath = "/api";
const produitPath = "/produit";
const parutionPath = "/parution";
const folioPath = "/folio";
const statusPath = "/status";
const alphaPath = oriPath + produitPath;

function makeListing() {
  // '/api/produit/ERR'
  apiList.push(oriPath + produitPath + errTest);
  for (let a = 0; a < productArr.length; a++) {
    for (let b = 0; b < parutionArr.length; b++) {
      //PARUTION
      // '/api/produit/ERR/parution'
      apiList.push(alphaPath + errTest + parutionPath);
      // '/api/produit/sli/parution/ERR'
      apiList.push(alphaPath + productArr[a] + parutionPath + errTest);
      // '/api/produit/ERR/parution/20160101'
      apiList.push(alphaPath + errTest + parutionPath + parutionArr[b]);
      // '/api/produit/ERR/parution/ERR'
      apiList.push(alphaPath + errTest + parutionPath + errTest);
      for (let c = 0; c < folioArr.length; c++) {
        //FOLIOS
        // '/api/produit/sli/parution/ERR/folio'
        apiList.push(alphaPath + productArr[a] + parutionPath + errTest + folioPath);
        // '/api/produit/ERR/parution/20160101/folio'
        apiList.push(alphaPath + errTest + parutionPath + parutionArr[b] + folioPath);
        // '/api/produit/ERR/parution/ERR/folio'
        apiList.push(alphaPath + errTest + parutionPath + errTest + folioPath);
        // PAGES
        // '/api/produit/sli/parution/20160101/folio/ERR'
        apiList.push(alphaPath + productArr[a] + parutionPath + parutionArr[b] + folioPath + errTest);
        // '/api/produit/ERR/parution/20160101/folio/ERR'
        apiList.push(alphaPath + errTest + parutionPath + parutionArr[b] + folioPath + errTest);
        // '/api/produit/ERR/parution/20160101/folio/01'
        apiList.push(alphaPath + errTest + parutionPath + parutionArr[b] + folioPath + folioArr[c]);
        // '/api/produit/sli/parution/ERR/folio/ERR'
        apiList.push(alphaPath + productArr[a] + parutionPath + errTest + folioPath + errTest);
        // '/api/produit/ERR/parution/ERR/folio/ERR'
        apiList.push(alphaPath + errTest + parutionPath + errTest + folioPath + errTest);
        // '/api/produit/sli/parution/ERR/folio/01'
        apiList.push(alphaPath + productArr[a] + parutionPath + errTest + folioPath + folioArr[c]);
        // '/api/produit/ERR/parution/ERR/folio/01'
        apiList.push(alphaPath + errTest + parutionPath + errTest + folioPath + folioArr[c]);
      }
    }
  }
}
makeListing();
module.exports = apiList;


Gruntfile.js
"use strict";
/*jslint node: true */
/*jshint esversion: 6 */
module.exports = function (grunt) {
  // Project configuration.
  grunt.initConfig({
    pkg: grunt.file.readJSON('package.json'),
    // dist/<%= pkg.srcfile %>./js
    jshint: {
      files: ['Gruntfile.js', 'src/server.js', 'src/provider/data.js', 'src/lib/lib.js', 'src/cluster.js'],
      options: {
        // options here to override JSHint defaults
        node: true,
        esversion: 6,
        globals: {
          console: true,
          module: true,
          document: true,
        }
      }
    },
    uglify: {
      options: {
        banner: '/*! <%= pkg.name %> - v<%= pkg.version %> - ' + '<%= grunt.template.today("yyyy-mm-dd") %> */',
        screwIE8: true,
        compress: {
          drop_console: true
        }
      },
      my_target: {
        files: [{
          expand: true,
          cwd: 'build/',
          src: '**/*.js',
          dest: 'dist/'
        }]
      },
    },
    watch: {
      files: ['src/server.js'],
      tasks: ['default']
    },
    babel: {
      options: {
        presets: ['es2015'],
        sourceMap: false
      },
      dist: {
        files: [{
          expand: true,
          cwd: "src/",
          src: ["**/*.js"],
          dest: "build/",
          ext: ".js"
        }]
      }
    },
    clean: ["build/", "dist/", "docs/"]
    // clean: ["build/", "dist/", "docs/", "coverage/"]
  });
  grunt.loadNpmTasks('grunt-babel');
  grunt.loadNpmTasks('grunt-contrib-clean');
  grunt.loadNpmTasks('grunt-contrib-uglify');
  grunt.loadNpmTasks('grunt-contrib-jshint');
  grunt.loadNpmTasks('grunt-contrib-watch');
  // Default task(s).
  grunt.registerTask('default', ['clean', 'jshint', 'babel', 'uglify']);
};


src/server.js
// INIT FILE SETUP
// =============================================================================
// - use strict option
"use strict";
// - jslint option
/*jslint node: true */
/*jshint esversion: 6 */
// REQUIRE
// =============================================================================
// - call the packages we need
// - must have to go !
const express = require('express');
const app = express();
const bodyParser = require('body-parser');
const dataFile = require('./provider/data');
const lib = require('./lib/lib');
const data = dataFile;
const port = process.env.PORT || 8e3;
const router = express.Router();
// - Configure `app` to use `bodyParser()`
app.use(bodyParser.urlencoded({
  extended: true
}));
// - Security disable x-powered-by
app.disable('x-powered-by');
// - This will let us get the data from a POST
app.use(bodyParser.json());
// - Configure `app.set()` to indent json with 2 spaces in httpResponse
app.set('json spaces', 2);
// FUNCTION
// =============================================================================
//     errEnd(err, res);
// return captured error to json httpResponse, status 400
//  1. set Http status to `400` (bad request) => `res.status(codeNumber);`
//  2. return a `json` with `err` => `res.json(err);``
/**
 * @method errEnd
 * @param  {any}   err [captured error]
 * @param  {array} res [captured http res]
 * @return {json}      [return http json]
 */
function errEnd(err, res) {
  res.status(400)
    .json({ "err": err });
}
// TODO: capture all put
// temp fonction
function capturePut(req, res) {
  console.log(req.params);
  res.json(req.params);
}
// API ROUTES
// =============================================================================
// - Here all the api route, get, post, put, delete
//
// /
// -----
// - `get`
router.route('/')
  .get((req, res) => {
    res.json({ message: 'Welcome on flatplan_api !' });
  });
// /produit/
// -----
// - `get`
router.route('/produit/')
  .get((req, res) => {
    res.json(data);
  });
// /produit/:produit
// -----
// - `get`
router.route('/produit/:produit')
  .get((req, res) => {
    lib.productIdx(req.params.produit)
      .then(productIdx => {
        res.json(data[productIdx]);
      })
      .catch(err => errEnd(err, res));
  })
  .put((req, res) => capturePut(req, res));
// /produit/:produit/parution/
// -----
// - `get`
router.route('/produit/:produit/parution/')
  .get((req, res) => {
    lib.productIdx(req.params.produit)
      .then(productIdx => {
        res.json(data[productIdx].parution);
      })
      .catch(err => errEnd(err, res));
  });
// /produit/:produit/parution/:parution
// -----
// - `get`
router.route('/produit/:produit/parution/:parution')
  .get((req, res) => {
    lib.productParution(req)
      .then(result => {
        res.json(data[result.productIdx].parution[result.parutionIdx]);
      })
      .catch(err => { errEnd(err, res); });
  })
  .put((req, res) => capturePut(req, res));
// /produit/:produit/parution/:parution/folio
// -----
// - `get`
router.route('/produit/:produit/parution/:parution/folio')
  .get((req, res) => {
    lib.productParution(req)
      .then(result => {
        res.json(data[result.productIdx].parution[result.parutionIdx].folio);
      })
      .catch(err => { errEnd(err, res); });
  });
// /produit/:produit/parution/:parution/folio/:folio
// -----
// TODO: find good wa
router.route('/produit/:produit/parution/:parution/folio/:folio')
  .get((req, res) => {
    lib.productParutionFolio(req)
      .then(result => {
        res.json(data[result.productIdx].parution[result.parutionIdx].folio[result.folioIdx]);
      })
      .catch(err => { errEnd(err, res); });
  })
  // - `put` update folio
  .put((req, res) => {
    lib.productParutionFolio(req)
      .then(result => {
        data[result.productIdx].parution[result.parutionIdx].folio[result.folioIdx].page = req.params.folio;
        res.json(data[result.productIdx].parution[result.parutionIdx].folio[result.folioIdx]);
      })
      .catch(err => { errEnd(err); });
  });
// /produit/:produit/parution/:parution/folio/:folio/status
// -----
// - `get` status of folio
router.route('/produit/:produit/parution/:parution/folio/:folio/status')
  .get((req, res) => {
    lib.productParutionFolio(req)
      .then(result => {
        res.json(data[result.productIdx].parution[result.parutionIdx].folio[result.folioIdx].status);
      })
      .catch(err => { errEnd(err, res); });
  });
// /produit/:produit/parution/:parution/folio/:folio/status/:status
// ------
// - `put` update status false === locked or true === unlocked
router.route('/produit/:produit/parution/:parution/folio/:folio/status/:status')
  .put((req, res) => {
    lib.productParutionFolio(req)
      .then(result => {
        data[result.productIdx].parution[result.parutionIdx].folio[result.folioIdx].status = req.params.status;
        res.json(data[result.productIdx].parution[result.parutionIdx].folio[result.folioIdx].status);
      })
      .catch(err => { errEnd(err); });
  });
// REGISTER ROUTE
// =============================================================================
// all routes will be prefixed with /api
app.use('/api', router);
// START THE SERVER
// =============================================================================
// express.listen port, up httpServer
app.listen(port);
console.log('Magic happens on port ' + port);
// DEBUG SUPPORT
// =============================================================================
module.exports = app;
//
//
//
//
//
//
//
//
//
// easter eggs
// router.route('/eggs', (req, res) => {
//  res.redirect('https://www.youtube.com/watch?v=3Yfy_QIgpsc');
// });


src/lib/lib.js
"use strict";
// BASE SETUP
// =============================================================================
// - Call the packages we need
const dataFile = require('../provider/data');
const data = dataFile;
// INIT LIB
// =============================================================================
const lib = [];
// FUNCTION
// =============================================================================
// productIdx
// ----------
//     productIdx(product);
// Return Promise find index from `product:string`
/**
 * [Return index product]
 * @method productIdx
 * @param  {string}   product [product]
 * @return {Promise}          [index product]
 */
lib.productIdx = product => {
  return new Promise((resolve, reject) => {
    const productIdx = data.map((el) => el.produit)
      .indexOf(product);
    if (productIdx < 0) {
      reject("lib.productIdx() => Product index not found ");
    }
    resolve(productIdx);
  });
};
// parutionIdx
// ----------
//     parutionIdx(productIdx, req);
// Return Promise find req.params.parution index from `productIdx:number`
/**
 * [Return index parution for index product]
 * @method parutionIdx
 * @param  {number}    productIdx [index product]
 * @param  {object}    req        [httprequest for inject0 req.params.parution]
 * @return {Promise}              [index parution]
 */
lib.parutionIdx = (productIdx, req) => {
  return new Promise((resolve, reject) => {
    const parutIdx = data[productIdx].parution.map((el) => el.parution)
      .indexOf(req.params.parution);
    if (parutIdx < 0) {
      reject("lib.parutionIdx() => Parution index not found ");
    }
    resolve(parutIdx);
  });
};
// folioIdx
// ----------
//     folioIdx(productIdx,parutionIdx, req);
// Return Promise find req.params.folio index from` `productIdx:number ` && `parutionIdx:number `
/**
 * [Return return index folio for index product & index parution]
 * @method folioIdx
 * @param  {number} productIdx  [index product]
 * @param  {number} parutionIdx [index parution]
 * @param  {object} req         [httprequest req.params.folio ]
 * @return {Promise}            [index folio]
 */
lib.folioIdx = (productIdx, parutionIdx, req) => {
  return new Promise((resolve, reject) => {
    const folioIdx = data[productIdx].parution[parutionIdx].folio.map((el) => el.page)
      .indexOf(req.params.folio);
    if (folioIdx < 0) {
      reject("lib.folioIdx() => folio index not found ");
    }
    resolve(folioIdx);
  });
};
// productParution
// ----------
//     productParution(req);
// Return Promise `{ 'productIdx': productIdx, 'parutionIdx': parutionIdx }`
/**
 * [productParution return object with Product et Parution index]
 * @method productParution
 * @param  {object}        req [request]
 * @return {object}            [{ 'productIdx': productIdx, 'parutionIdx': parutionIdx }]
 */
lib.productParution = (req) => {
  return new Promise((resolve, reject) => {
    lib.productIdx(req.params.produit)
      .then(productIdx => {
        lib.parutionIdx(productIdx, req)
          .then(parutionIdx => {
            resolve({ 'productIdx': productIdx, 'parutionIdx': parutionIdx });
          })
          .catch(err => reject(err));
      })
      .catch(err => reject(err));
  });
};
// productParutionFolio
// ----------
//     productParutionFolio(req);
// Return Promise `{ 'productIdx': productIdx, 'parutionIdx': parutionIdx, 'folioIdx': folioIdx }`
/**
 * [productParutionFolio return object with Product & Parution index & Folio index]
 * @method productParutionFolio
 * @param  {object}        req [request]
 * @return {object}            [{ 'productIdx': productIdx, 'parutionIdx': parutionIdx, 'folioIdx': folioIdx }]
 */
lib.productParutionFolio = (req) => {
  return new Promise((resolve, reject) => {
    lib.productIdx(req.params.produit)
      .then(productIdx => {
        lib.parutionIdx(productIdx, req)
          .then(parutionIdx => {
            lib.folioIdx(productIdx, parutionIdx, req)
              .then(folioIdx => {
                resolve({ 'productIdx': productIdx, 'parutionIdx': parutionIdx, 'folioIdx': folioIdx });
              })
              .catch(err => reject(err));
          })
          .catch(err => reject(err));
      })
      .catch(err => reject(err));
  });
};
// EXPORT LIB // DEBUG SUPPORT
// =============================================================================
module.exports = lib;
// if (require.resolve('mongoose')) reject('err : mongoose require ');


src/provider/data.js
'use strict';
let data = [];
module.exports = data = [{
  "produit": "ewe",
  "name": "Les Echos Weekend",
  "parution": [{
    "parution": "20160101",
    "folio": [{
      "page": "01",
      "status": "locked",
      "workflow": "bat"
    }, {
      "page": "02",
      "status": "locked",
      "workflow": "bat"
    }]
  }, {
    "parution": "20160102",
    "folio": [{
      "page": "01",
      "status": "locked",
      "workflow": "bat"
    }, {
      "page": "02",
      "status": "locked",
      "workflow": "bat"
    }]
  }]
}, {
  "produit": "sli",
  "name": "Serie Limitée",
  "parution": [{
    "parution": "20160101",
    "folio": [{
      "page": "01",
      "status": "locked",
      "workflow": "bat"
    }, {
      "page": "02",
      "status": "locked",
      "workflow": "bat"
    }]
  }, {
    "parution": "20160102",
    "folio": [{
      "page": "01",
      "status": "locked",
      "workflow": "bat"
    }, {
      "page": "02",
      "status": "locked",
      "workflow": "bat"
    }]
  }]
}];


src/1.js
// server.js
// BASE SETUP
// =============================================================================

// call the packages we need
var express = require('express'); // call express
var app = express(); // define our app using express
var bodyParser = require('body-parser');

// configure app to use bodyParser()
// this will let us get the data from a POST
app.use(bodyParser.urlencoded({ extended: true }));
app.use(bodyParser.json());
app.set('json spaces', 2);
var port = process.env.PORT || 8080; // set our port

// ROUTES FOR OUR API
// =============================================================================
var router = express.Router(); // get an instance of the express Router

var data = [
    {
        "produit": "ewe",
        "name": "Les Echos Weekend",
        "parution": [{
            "parution": "20160101",
            "folio": [{
                "page": 01,
                "status": "locked",
                "workflow": "bat"
            }, {
                "page": 02,
                "status": "locked",
                "workflow": "bat"
            }]
        }, {
            "parution": "20160102",
            "folio": [{
                "page": 01,
                "status": "locked",
                "workflow": "bat"
            }, {
                "page": 02,
                "status": "locked",
                "workflow": "bat"
            }]
        }]
    }, {
        "produit": "sli",
        "name": "Serie Limitée",
        "parution": [{
            "parution": "20160101",
            "folio": [{
                "page": 01,
                "status": "locked",
                "workflow": "bat"
            }, {
                "page": 02,
                "status": "locked",
                "workflow": "bat"
            }]
        }, {
            "parution": "20160102",
            "folio": [{
                "page": 01,
                "status": "locked",
                "workflow": "bat"
            }, {
                "page": 02,
                "status": "locked",
                "workflow": "bat"
            }]
        }]
    }
];



// test route to make sure everything is working (accessed at GET http://localhost:8080/api)
router.get('/', function(req, res) {
    // res.json({ message: 'Bienvenue sur l"api de pierre !' });
    res.redirect('https://www.youtube.com/watch?v=3Yfy_QIgpsc')
});

router.get('/produit/', function(req, res) {
    res.json(data);
});

router.get('/produit/:produit', function(req, res) {
    var idx = data.map((el) => el.produit).indexOf(req.params.produit);
    res.json(data[idx]);
});

router.get('/produit/:produit/parution/', function(req, res) {
    var idx = data.map((el) => el.produit).indexOf(req.params.produit);
    // var idxparution= data[idxPro].map((el) => el.parution).indexOf(req.params.produit);
    res.json(data[idx].parution);
});

router.get('/produit/:produit/parution/:parution', function(req, res) {
    console.log(req.params)
    var idx = data.map((el) => el.produit).indexOf(req.params.produit);
    var idxparution = data[idx].parution.map((el) => el.parution).indexOf(req.params.parution);
    console.log(idxparution)
    res.json(data[idx].parution[idxparution]);
});

router.get('/produit/:produit/parution/:parution/folio', function(req, res) {
    console.log(req.params)
    var idx = data.map((el) => el.produit).indexOf(req.params.produit);
    var idxparution = data[idx].parution.map((el) => el.parution).indexOf(req.params.parution);
    console.log(idxparution)
    res.json(data[idx].parution[idxparution].folio);
});
router.get('/produit/:produit/parution/:parution/folio/:folio', function(req, res) {
    console.log(req.params)
    var idx = data.map((el) => el.produit).indexOf(req.params.produit);
    var idxparution = data[idx].parution.map((el) => el.parution).indexOf(req.params.parution);
    var idxFolio = data[idx].parution[idxparution].folio.map((el) => el.page).indexOf(Number(req.params.folio));
    console.log(idxFolio)
    res.json(data[idx].parution[idxparution].folio[idxFolio]);
});
router.get('/produit/:produit/parution/:parution/folio/:folio/status', function(req, res) {
    console.log(req.params)
    var idx = data.map((el) => el.produit).indexOf(req.params.produit);
    var idxparution = data[idx].parution.map((el) => el.parution).indexOf(req.params.parution);
    var idxFolio = data[idx].parution[idxparution].folio.map((el) => el.page).indexOf(Number(req.params.folio));
    console.log(idxFolio)
    res.json(data[idx].parution[idxparution].folio[idxFolio].status);
});

// more routes for our API will happen here

// REGISTER OUR ROUTES -------------------------------
// all of our routes will be prefixed with /api
app.use('/api', router);

// START THE SERVER
// =============================================================================
app.listen(port);
console.log('Magic happens on port ' + port);
